core/unix/signal.c

```diff
@@ -3235,20 +3235,10 @@ thread_set_self_context(void *cxt, bool is_detach_external)
     ASSERT_NOT_IMPLEMENTED(false); /* PR 405694: can't use regular sigreturn! */
 #endif
     memset(&frame, 0, sizeof(frame));
-#if defined(X86)
-    dcontext_t *dcontext = get_thread_private_dcontext();
-#endif
 #ifdef LINUX
-#    ifdef X86
-    byte *xstate = get_and_initialize_xstate_buffer(dcontext);
-    frame.uc.uc_mcontext.fpstate = &((kernel_xstate_t *)xstate)->fpstate;
-#    endif /* X86 */
     frame.uc.uc_mcontext = *sc;
 #endif
     IF_ARM(ASSERT_NOT_TESTED());
-#if defined(X86)
-    save_fpstate(dcontext, &frame);
```

Comment:
I'm not sure removing this is safe. @abhinav92003 PTAL.

Reply:
fwiw, save_fpstate does nothing if fpstate is NULL (which it is here, due to line 3246 undoing 3244). so removing it should not break anything that was not already broken

Reply:
Oh, maybe it's needed for some reason on non-`LINUX` platforms?

Reply:
The current code is indeed confusing.  There is the matter of the x87 floating point state, which is not preserved in DR's mcontext: that's where we need to get state from the underlying processor.  That may need fxsave into a separate buffer and then copy just the x87 stuff over via convert_fxsave_to_fpstate()?  If that's too much for your PR, since like you said it was already broken, at least add a TODO on doing that and add a note to the issue that x87 state is still broken?  These days x87 usage is pretty rare so if you want to not tackle preserving it or adding a test (we don't have many tests of it for other things either, to be fair) just adding a TODO seems ok.

For non-LINUX: today that would just be MACOS since Android ends up defining LINUX (yes, surprising) -- OK to not worry about it for this PR: it's already NYI (assert on line 3274).

Reply:
I think the current save_fpstate code actually might be fine: it only appears to copy the x87 part of the state, which is exactly what we want. I put it back and it still passes my test.

Reply:
Right, it does fxsave, not xsave: so just x87: agreed.

Reply:
> Right, it does fxsave, not xsave: so just x87: agreed.

So the bottom of save_fpstate() calls save_xmm() which gets xgetbv from the processor and pulls the XMM/YMM/ZMM from mcontext.  Is that XMM/YMM/ZMM redundant with the mcontext_to_sigcontext() that now has an fpstate in thread_set_self_mcontext()?

Reply:
ono, you are right. it seems like the only change we needed in the first place is to actually initialize fpstate before calling save_fpstate. Sorry for the confusion around the different fp states :/

Reply:
The code looks simpler and better now.  And if the new test passes then presumably nothing is broken with the simpler code; sounds like a win.